--- ./cvsspam.rb~	2005-05-14 23:28:39.000000000 +0300
+++ ./cvsspam.rb	2005-05-19 13:19:50.000000000 +0300
@@ -1444,7 +1444,7 @@
   tr.head {border-bottom-width:1px;border-bottom-style:solid;}
   tr.head td {padding:0;padding-top:.2em;}
   .task {background-color:#ffff00;}
-  .comment {padding:4px;border:1px dashed #000000;background-color:#ffffdd}
+  .comment {white-space:-moz-pre-wrap;white-space:-pre-wrap;white-space:-o-pre-wrap;white-space:pre-wrap;word-wrap:break-word;padding:4px;border:1px dashed #000000;background-color:#ffffdd}
   .error {color:red;}
   hr {border-width:0px;height:2px;background:black;}
 --></style>
