--- cvsspam-0.2.12/svn_post_commit_hook.rb	2009-03-04 22:21:45.091026513 +0200
+++ cvsspam-0.2.12-svn/svn_post_commit_hook.rb	2009-03-04 22:21:30.394359128 +0200
@@ -28,12 +28,14 @@
 end
 
 def cleanup
-  File.unlink("#{$datadir}/logfile")
-  Dir.rmdir($datadir)
+  unless $debug
+	  File.unlink("#{$datadir}/logfile")
+	  Dir.rmdir($datadir)
+  end
 end
 
 def send_email
-  cmd = File.dirname($0) + "/cvsspam.rb"
+  cmd = File.dirname($0) + "/svn_cvsspam.rb"
   unless system(cmd,"--svn","#{$datadir}/logfile", *$passthrough_args)
     fail "problem running '#{cmd}'"
   end
