--- collect_diffs.rb.org	2004-12-11 21:38:10.000000000 +0100
+++ collect_diffs.rb	2004-12-11 21:46:35.235302401 +0100
@@ -100,7 +100,7 @@
 def collect_antique_style_args(cvs_info)
   # remove leading slashes that may appear due to the user entering trailing
   # slashes in their CVSROOT specification
-  cvs_info.sub!(/^\/+/, "")
+  cvs_info = cvs_info.sub(/^\/+/, "")
 
   unless cvs_info.slice(0, $repository_path.length+1) == "#{$repository_path} "
     fail "calculated repository path ('#{$repository_path}') doesn't match start of command line arg ('#{cvs_info}')"
