--- cvsspam/svn_post_commit_hook.rb~	2009-03-04 22:22:15.000000000 +0200
+++ cvsspam/svn_post_commit_hook.rb	2009-03-04 23:07:24.382496937 +0200
@@ -88,8 +88,6 @@
   unless FileTest.directory?($repository)
     usage("no such directory: #{$repository.inspect}")
   end
-  $repository =~ /([^\/]+$)/ 
-  $shortrepo = $1
 end
 
 # runs the given svnlook subcommand
@@ -173,7 +171,7 @@
   prev_rev= $revision-1
   next_rev= $revision
   out.puts "#V #{prev_rev},#{next_rev}"
-  out.puts "#M #{$shortrepo}/#{path}"
+  out.puts "#M #{path}"
   out.puts "#U diff x x"
   out.puts "#U Binary files x and y differ"
 end
@@ -187,7 +185,7 @@
   next_rev = m[1].to_i
   diff2 = lines.current
   out.puts "#V #{prev_rev},#{next_rev}"
-  out.puts "#M #{$shortrepo}/#{path}"
+  out.puts "#M #{path}"
   out.puts "#U #{diff1}"
   out.puts "#U #{diff2}"
   while lines.next_line && lines.current =~ /^[-\+ @\\]/
@@ -208,7 +206,7 @@
 def process_added_binary_diff(out, lines, path)
   next_rev= $revision
   out.puts "#V NONE,#{next_rev}"
-  out.puts "#A #{$shortrepo}/#{path}"
+  out.puts "#A #{path}"
   out.puts "#U diff x x"
   out.puts "#U Binary file x added"
 end
@@ -221,7 +219,7 @@
   next_rev = m[1].to_i
   diff2 = lines.current
   out.puts "#V NONE,#{next_rev}"
-  out.puts "#A #{$shortrepo}/#{path}"
+  out.puts "#A #{path}"
   out.puts "#U #{diff1}"
   out.puts "#U #{diff2}"
   while lines.next_line && lines.current =~ /^[-\+ @\\]/
@@ -238,7 +236,7 @@
   next_rev = m[1].to_i
   diff2 = lines.current
   out.puts "#V #{prev_rev},NONE"
-  out.puts "#R #{$shortrepo}/#{path}"
+  out.puts "#R #{path}"
   out.puts "#U #{diff1}"
   out.puts "#U #{diff2}"
   while lines.next_line && lines.current =~ /^[-\+ @\\]/
@@ -287,7 +285,7 @@
   lines.assert_next(/^_+$/)
   return unless lines.next_line
   out.puts "#V #{prev_rev},#{next_rev}"
-  out.puts "#P #{$shortrepo}/#{path}"
+  out.puts "#P #{path}"
 # The first three get consumed and not highlighted
   out.puts "#U "
   out.puts "#U Property changes:"
@@ -316,8 +314,8 @@
 def handle_copy(out, lines, path, from_ref, from_file)
   prev_rev= $revision-1
   next_rev= $revision
-  out.puts "#V #{$shortrepo}/#{from_file}:#{prev_rev},#{next_rev}"
-  out.puts "#C #{$shortrepo}/#{path}"
+  out.puts "#V #{from_file}:#{prev_rev},#{next_rev}"
+  out.puts "#C #{path}"
   if lines.next_line && lines.current =~ /^=+$/
     m = lines.assert_next(/^---.*\(rev (\d+)\)$/)
     prev_rev = m[1].to_i
@@ -332,7 +330,7 @@
     end
   else
     out.puts "#U "
-    out.puts "#U Copied from #{$shortrepo}/#{from_file}:#{from_ref}"
+    out.puts "#U Copied from #{from_file}:#{from_ref}"
     out.puts "#U "
   end
 end
