--- ./cvsspam.rb	2004-12-10 01:51:21.000000000 +0200
+++ ./cvsspam.rb	2005-02-15 21:46:58.000000000 +0200
@@ -1568,9 +1568,9 @@
   if File.exists?($users_file)
     File.open($users_file) do |io|
       io.each_line do |line|
-        if line =~ /^([^:]+)\s*:\s*([^\n\r]+)/
+        if line =~ /^([^:]+)\s*:\s*(['"]?)([^\n\r]+)(\2)/
           if address == $1
-            return $2
+            return $3
           end
         end
       end
