--- cvsspam/cvsspam.rb~	2006-02-28 11:30:01.000000000 +0200
+++ cvsspam/cvsspam.rb	2006-03-02 00:49:48.000000000 +0200
@@ -546,6 +546,10 @@
   match =~ /([0-9]+)/
   "<a href=\"#{$ticketURL.sub(/%s/, $1)}\">#{match}</a>"
 }
+issueSub = proc { |match|
+  match =~ /([0-9]+)/
+  "<a href=\"#{$issueURL.sub(/%s/, $1)}\">#{match}</a>"
+}
 wikiSub = proc { |match| 
   match =~ /\[\[(.*)\]\]/
   raw = $1
@@ -1281,6 +1285,7 @@
 $wikiURL = nil
 $jiraURL = nil
 $ticketURL = nil
+$issueURL = nil
 $viewcvsURL = nil
 $choraURL = nil
 $cvswebURL = nil
@@ -1393,6 +1398,9 @@
 if $ticketURL != nil
   commentSubstitutions['\b[Tt][Ii][Cc][Kk][Ee][Tt]\s*#?[0-9]+\b'] = ticketSub
 end
+if $issueURL != nil
+  commentSubstitutions['\b[Ii][Ss][Ss][Uu][Ee]\s*#?[0-9]+\b'] = issueSub
+end
 if $wikiURL != nil
   commentSubstitutions['\[\[.+\]\]'] = wikiSub
 end
