--- cvsspam-0.2.12/cvsspam.rb~	2010-04-07 18:25:49.252489451 +0300
+++ cvsspam-0.2.12/cvsspam.rb	2010-04-07 18:27:22.000000000 +0300
@@ -169,7 +171,7 @@
 
   # test to see of the given string contains non-ASCII characters
   def requires_rfc2047?(word)
-    (word =~ /[\177-\377]/) != nil
+    (word.force_encoding('ASCII-8BIT') =~ /[\177-\377]/) != nil
   end
 end
 
@@ -2242,6 +2244,7 @@
   if File.exists?($users_file)
     File.open($users_file) do |io|
       io.each_line do |line|
+        line.force_encoding($users_file_charset)
         if line =~ /^([^:]+)\s*:\s*(['"]?)([^\n\r]+)(\2)/
           if email.address == $1
             return EmailAddress.new($3, $users_file_charset)
